### PR TITLE
fix memory leak

### PR DESCRIPTION
**Pull request by rabauke**
Fixes a memory leak and other issues.

### CODE REVIEW

File: src/libs/metadata.c [hunk 1]
```diff
@@ -736,23 +736,27 @@ void *legacy_params(dt_lib_module_t *self, const void *const old_params, const s
...
     size_t new_params_size = old_params_size + 1;
-    void *new_params = calloc(sizeof(char), new_params_size);
+    char *new_params = calloc(sizeof(char), new_params_size);
 
-    char *buf = (char *)old_params;
+    const char *buf = (const char *)old_params;
 
     // <title>\0<description>\0<rights>\0<creator>\0<publisher>
-    char *metadata[DT_METADATA_NUMBER];
-    int32_t metadata_len[DT_METADATA_NUMBER];
```
Comment (rabauke) [inline src/libs/metadata.c]: Elements of `metadata_len` are used as `size_t` arguments to `memcpy`.

File: src/libs/metadata.c [hunk 2]
```diff
@@ -736,23 +736,27 @@ void *legacy_params(dt_lib_module_t *self, const void *const old_params, const s
...
+    const char *metadata[DT_METADATA_NUMBER];
+    size_t metadata_len[DT_METADATA_NUMBER];
     for(unsigned int i = 0; i < DT_METADATA_NUMBER; i++)
     {
       metadata[i] = buf;
-      if(!metadata[i]) return NULL;
```
Comment (rabauke) [inline src/libs/metadata.c]: Here the memory leak occurs.

File: src/libs/metadata.c [hunk 3]
```diff
@@ -736,23 +736,27 @@ void *legacy_params(dt_lib_module_t *self, const void *const old_params, const s
...
+        return NULL;
+      }
       metadata_len[i] = strlen(metadata[i]) + 1;
       buf += metadata_len[i];
     }
 
     // <creator>\0<publisher>\0<title>\0<description>\0<rights>
-    int pos = 0;
+    size_t pos = 0;
     memcpy(new_params + pos, metadata[3], metadata_len[3]);
```
Comment (rabauke) [inline src/libs/metadata.c]: Pointer arithmetic with pointer to void is not legal C. It's a language extension of the [GNU compiler](https://gcc.gnu.org/onlinedocs/gcc/Pointer-Arith.html).

File: src/libs/metadata.c [hunk 4]
```diff
@@ -736,23 +736,27 @@ void *legacy_params(dt_lib_module_t *self, const void *const old_params, const s
   if(old_version == 1)
   {
     size_t new_params_size = old_params_size + 1;
-    void *new_params = calloc(sizeof(char), new_params_size);
+    char *new_params = calloc(sizeof(char), new_params_size);
 
-    char *buf = (char *)old_params;
```
Comment (rabauke) [inline src/libs/metadata.c]: Do not cast away constness!

File: src/libs/metadata.c [hunk 5]
```diff
@@ -736,23 +736,27 @@ void *legacy_params(dt_lib_module_t *self, const void *const old_params, const s
   if(old_version == 1)
   {
     size_t new_params_size = old_params_size + 1;
-    void *new_params = calloc(sizeof(char), new_params_size);
```
Comment (rabauke) [inline src/libs/metadata.c]: Must be pointer to `char` as we are going to deal with chars.

**Comment by TurboGit** [comment]
Thanks a lot for the fix and detailed analysis !

**Pull request merged**